Implement quality selection for transcoded videos

ktprograms opened the pull request:
## PR Checklist

<!-- For the checkbox formatting to work properly, make sure there are no spaces on either side of the "x" -->

Please check all that apply to this PR using "x":

- [x] I have checked that this PR is not a duplicate of an existing PR (open, closed or merged)
- [x] I have checked that this PR does not introduce a breaking change

## PR Type

What kind of change does this PR introduce?

- [x] Feature

## Fixes

Issue Number: #113 

## What is the current behavior?

No quality selector

## What is the new behavior?

Quality selector next to the speed selector.

- An `Auto` selection to let Exoplayer decide the best quality to use (also for non-transcoded videos)
- Settings for default qualities on Wi-Fi or Mobile Data

| Player view | Settings |
|--|--|
| <img height="500" alt="Player view" src="https://user-images.githubusercontent.com/71804605/159434445-584f730a-167a-4932-92ec-3193563ec341.png"/> | <img height="500" alt="Settings" src="https://user-images.githubusercontent.com/71804605/159434603-d7dd05da-3a31-4f18-863f-714cd1efe120.png"/> |

## Note

The Exoplayer dependency is bumped to 2.16.0 to allow using the new track selection APIs. To the best of my knowledge this didn't break anything, but I didn't test live streams.

ktprograms commented on app/src/main/java/com/odysee/app/ui/findcontent/FileViewFragment.java:
```diff
@@ -3565,15 +3751,6 @@ private void checkOwnClaim() {
     }
 
     public static class StreamLoadErrorPolicy extends DefaultLoadErrorHandlingPolicy {
-        @Override
-        public long getRetryDelayMsFor(int dataType, long loadDurationMs, IOException exception, int errorCount) {
-            return exception instanceof ParserException
-                    || exception instanceof FileNotFoundException
-                    || exception instanceof Loader.UnexpectedLoaderException
-                    ? C.TIME_UNSET
-                    : Math.min((errorCount - 1) * 1000, 5000);
-        }
-
```
This is removed because it is not compatible with Exoplayer 2.16.0, but to the best of my knowledge this was implementing the default behaviour for this method.

akinwale replied:
This function is a custom error handler meant to keep trying to load the stream every 5 seconds if we didn't encounter one of the following fatal exceptions (resulting in an unrecoverable failure): `ParserException`, `FileNotFoundException` or `Loader.UnexpectedLoaderException`. If this is the default behaviour in Exoplayer 2.16.0, that's fine, otherwise we have to find an alternative way to implement this.

ktprograms replied:
```java
  /**
   * Retries for any exception that is not a subclass of {@link ParserException}, {@link
   * FileNotFoundException}, {@link CleartextNotPermittedException} or {@link
   * UnexpectedLoaderException}, and for which {@link
   * DataSourceException#isCausedByPositionOutOfRange} returns {@code false}. The retry delay is
   * calculated as {@code Math.min((errorCount - 1) * 1000, 5000)}.
   */
  @Override
  public long getRetryDelayMsFor(LoadErrorInfo loadErrorInfo) {
    IOException exception = loadErrorInfo.exception;
    return exception instanceof ParserException
            || exception instanceof FileNotFoundException
            || exception instanceof CleartextNotPermittedException
            || exception instanceof UnexpectedLoaderException
            || DataSourceException.isCausedByPositionOutOfRange(exception)
        ? C.TIME_UNSET
        : min((loadErrorInfo.errorCount - 1) * 1000, 5000);
  }
```

[This](https://github.com/google/ExoPlayer/blob/9f3868d6a07ba4baf1b73aa1239800f72ab9dacc/library/core/src/main/java/com/google/android/exoplayer2/upstream/DefaultLoadErrorHandlingPolicy.java#L101-L118) is the default implementation of this method, it considers `CleartextNotPermittedException` and `DataSourceException.isCausedByPositionOutOfRange(exception)` as non-retryable errors.

Do you think this is ok, or should the previous functionality be used?

kekkyojin commented on app/src/main/java/com/odysee/app/ui/findcontent/FileViewFragment.java:
```diff
@@ -1903,14 +1943,70 @@ private void playMedia() {
-                    MainActivity.appPlayer.prepare();
+    private void getStreamingUrlAndInitializePlayer() {
+        ExecutorService executor = Executors.newSingleThreadExecutor();
+        executor.execute(() -> {
```
Executor must be shutdown()

kekkyojin commented:
Also, Odysee is looking for buffering events when playing. Maybe changing the quality will generate a new buffering event, and as position may be different from the start of the stream, the event might be reported to analytics backend, which would probably be a false positive, as buffering could be required in order to change to the new quality and it has been because a user initiated action.

ktprograms commented:
Yes, I think I does buffer when changing quality, any idea how to prevent it sending false positives?

ktprograms commented on app/src/main/java/com/odysee/app/ui/findcontent/FileViewFragment.java:
```diff
@@ -1901,14 +1940,71 @@ public void onPlayWhenReadyChanged(boolean playWhenReady, int reason) {
+                ex.printStackTrace();
             }
+        });
+        executor.shutdown();
```
Like this?

kekkyojin replied:
There are different ways to do background work on Java. I normally create a thread and I use the Future API to run the code. That allows me to shutdown the executor when the task has finished, as the background thread will be blocked while the UI thread will not be.

You could left it as it is now, if it is not crashing the app or try to refactor the code to use a Future.

shutdown() wait for tasks which were started already -submitted AND started-; if the task was not started, your code will not run as the executor will be shutdown.

Executors are kept on memory. That's why they must be shutdown. Odysee have had them without shutting them down and it worked fine. This is more a performance improvement.

ktprograms replied:
It seems to be working fine, so I'll just leave it as it is.

akinwale reviewed (changes requested):
Looks great, but a few minor changes:
* Make the speed selector container smaller, and the quality selector container a little wider, with an increased left padding, as it appears to be to close to the duration bar. This becomes more apparent when you select `1080p`.
* Reset the quality selector display to `Auto` every time the user launches a new video. I selected `144p` on a video, and all subsequent videos I tried to play ended up displaying `144p`, even with `144p` not being one of the available options.
* Remove the getter and setter code for `permanentUrl`.

This isn't a blocker, but I was curious if Exoplayer automatically detects whether the device is connected via WiFi or mobile data, or if we have to write code for this functionality. If it's the latter, we can create a new issue as that is outside the scope for this issue. Thanks.

kekkyojin commented on app/src/main/java/com/odysee/app/model/Claim.java:
```diff
@@ -480,6 +484,10 @@ public void setDisliked(boolean d) {
+    public void setPermanentUrl(String p) {
+        this.permanentUrl = p;
+    }
+
```
This code allows this to be compiled on any Java IDE while offering exactly what is needed.

getters and setters are simple methods which doesn’t take a lot of time to developers to write. Even IDEs add a way to implement them without the need of any external tool being installed.

akinwale replied:
The Gradle dependency has already been added which lets the code compile without any errors. This makes the additional code redundant, unless the getter / setter is being implemented with additional functionality.

kekkyojin commented:
> Yes, I think I does buffer when changing quality, any idea how to prevent it sending false positives?

Add a global boolean variable set to true when user changes the quality and reset it after player starts playing with the new quality setting. Then on the buffering event listener, check to see if the event was triggered when changing to the new quality setting

ktprograms commented:
> This isn't a blocker, but I was curious if Exoplayer automatically detects whether the device is connected via WiFi or mobile data, or if we have to write code for this functionality. If it's the latter, we can create a new issue as that is outside the scope for this issue. Thanks.

@akinwale I don't quite get what you mean. What functionality should be performed when connected via Wi-Fi or mobile data?

ktprograms commented:
That's what https://github.com/OdyseeTeam/odysee-android/pull/155/files#diff-877c3bf19aa98ed68e9e8f93236a2ab1bcf1e25b69dae15739f293ee84e7df95R2134-R2140 is meant to do, although now that I look at [how NewPipe does it](https://github.com/TeamNewPipe/NewPipe/blob/5459a55406ae72783584f84c1a8410e10903ba8a/app/src/main/java/org/schabi/newpipe/util/ListHelper.java#L558-L566), I think that's a better way.

ktprograms commented:
Changes made in latest push:
- Removed setter and getter for `permanentUrl`.
- Fixed the logic in the quality selector not considering qualities lower than available, or non-transcoded videos.
- Changed the sizes and alignment of the speed and quality containers to add more space after the duration bar.
- Use the metered network detection code from NewPipe.
- Don't report buffer events when buffering due to changing quality.

akinwale reviewed (approved):
Everything looks good. Thanks.

ktprograms commented:
Thanks!
